### PR TITLE
Added verify command

### LINKED ISSUE #69: Integrity test an repository at any time
Add a cli-command to verify your repository integrity at any time instead of just when restoring an repo.

### PR DESCRIPTION
**Pull request by craftamap**
Closes #69

### CODE REVIEW

File: verify_test.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+			if err != nil {
+				t.Errorf("Failed saving chunk-index: %s", err)
+				return
+			}
+
+		}
+
+		tt.ErrorFunction(dir)
+		if err != nil {
+			t.Errorf("Failed doing destrutive function: %s", err)
```
Comment (muesli) [inline verify_test.go:139]: Typo in `destructive`.

File: verify_test.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+				t.Errorf("Failed to verify snapshot: %s", err)
+			}
+			errors := make([]error, 0)
+			for p := range progress {
+				if p.Error != nil {
+					errors = append(errors, p.Error)
+				}
+			}
+			if len(errors) != tt.NumberOfExpectedErrors {
+				t.Errorf("%d errors where expected but %d occured", tt.NumberOfExpectedErrors, len(errors))
```
Comment (muesli) [inline verify_test.go:161]: Typo in `were`.

File: verify_test.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
```
Comment (muesli) [inline verify_test.go:267]: Typo in `were`.

File: verify_test.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
```
Comment (muesli) [inline verify_test.go:373]: Typo in `were`.

**Comment by craftamap** [comment]
@muesli as you might have seen, I resolved my typos :)

File: cmd/knoxite/verify.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		progress, err := knoxite.VerifyRepo(repository, opts.Percentage)
+		if err != nil {
+			errors = append(errors, err)
+			return err
+		}
+
+		pb := &goprogressbar.ProgressBar{Total: 1000, Width: 40}
+		lastPath := ""
+	
+
```
Comment (muesli) [inline cmd/knoxite/verify.go:64]: Let's remove redundant newlines.

File: cmd/knoxite/verify.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+			pb.LazyPrint()
+
+		}
+		fmt.Println()
+		fmt.Printf("Verify done: %d errors\n", len(errors))
+		return nil
+	}
+	return err
+
```
Comment (muesli) [inline cmd/knoxite/verify.go:94]: Let's remove redundant newlines.

File: cmd/knoxite/verify.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	if err == nil {
+		progress, err := knoxite.VerifyVolume(repository, volumeId, opts.Percentage)
+		if err != nil {
+			errors = append(errors, err)
+			return err
+		}
+
+		pb := &goprogressbar.ProgressBar{Total: 1000, Width: 40}
+		lastPath := ""
+
```
Comment (muesli) [inline cmd/knoxite/verify.go:109]: Let's remove redundant newlines.

File: cmd/knoxite/verify.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	if err == nil {
+		progress, err := knoxite.VerifySnapshot(repository, snapshotId, opts.Percentage)
+		if err != nil {
+			errors = append(errors, err)
+			return err
+		}
+
+		pb := &goprogressbar.ProgressBar{Total: 1000, Width: 40}
+		lastPath := ""
+
```
Comment (muesli) [inline cmd/knoxite/verify.go:154]: Let's remove redundant newlines.

File: verify.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+import (
+	"math"
+	"math/rand"
+)
+
+func VerifyRepo(repository Repository, Percentage int) (prog chan Progress, err error) {
+	prog = make(chan Progress)
+
+	go func() {
+
```
Comment (muesli) [inline verify.go:19]: Let's remove redundant newlines.

File: verify.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+			}
+		}
+
+		archives := make([]string, 0)
+		for archiveHash := range archiveToSnapshot {
+			archives = append(archives, archiveHash)
+		}
+
+		// get all keys of the snapshot Archives
+
```
Comment (muesli) [inline verify.go:42]: Let's remove redundant newlines.

File: verify.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+		for _, snapshotHash := range volume.Snapshots {
+			_, snapshot, err := repository.FindSnapshot(snapshotHash)
+			if err != nil {
+				prog <- newProgressError(err)
+			}
+
+			for archiveHash := range snapshot.Archives {
+				archiveToSnapshot[archiveHash] = snapshot
+			}
+
```
Comment (muesli) [inline verify.go:99]: Let's remove redundant newlines.

File: verify.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+		}
+
+		archives := make([]string, 0)
+		for archiveHash := range archiveToSnapshot {
+			archives = append(archives, archiveHash)
+		}
+
+		// get all keys of the snapshot Archives
+
```
Comment (muesli) [inline verify.go:108]: Let's remove redundant newlines.

File: verify.go [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+			prog <- p
+
+			err := VerifyArchive(repository, *snapshot.Archives[archiveKey])
+			if err != nil {
+				prog <- newProgressError(err)
+			}
+
+			p.CurrentItemStats.Transferred += uint64((*snapshot.Archives[archiveKey]).Size)
+			prog <- p
+
```
Comment (muesli) [inline verify.go:137]: Let's remove redundant newlines.

File: verify.go [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+			if erri != nil {
+				return erri
+			}
+
+			chunk := arc.Chunks[idx]
+			_, errc := loadChunk(repository, arc, chunk)
+			if errc != nil {
+				return errc
+			}
+
```
Comment (muesli) [inline verify.go:209]: Let's remove redundant newlines.

File: verify.go [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+		}
+		return nil
+	}
+	return nil
+
```
Comment (muesli) [inline verify.go:214]: Let's remove redundant newlines.

File: verify_test.go [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+		return nil
+
+	}, 1, 100},
+}
+
+func TestVerifyRepo(t *testing.T) {
+	testPassword := "this_is_a_password"
+
+	for _, tt := range verifyTestCases {
+
```
Comment (muesli) [inline verify_test.go:69]: Let's remove redundant newlines.

File: verify_test.go [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+			if err != nil {
+				t.Errorf("Failed saving volume: %s", err)
+				return
+			}
+			err = index.Save(&r)
+			if err != nil {
+				t.Errorf("Failed saving chunk-index: %s", err)
+				return
+			}
+
```
Comment (muesli) [inline verify_test.go:134]: Let's remove redundant newlines.

File: verify_test.go [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+			for p := range progress {
+				if p.Error != nil {
+					errors = append(errors, p.Error)
+				}
+			}
+			if len(errors) != tt.NumberOfExpectedErrors {
+				t.Errorf("%d errors were expected but %d occured", tt.NumberOfExpectedErrors, len(errors))
+			}
+		}
+
```
Comment (muesli) [inline verify_test.go:164]: Let's remove redundant newlines.

File: verify_test.go [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+}
+
+func TestVerifyVolume(t *testing.T) {
+	testPassword := "this_is_a_password"
+
+	for _, tt := range verifyTestCases {
+
```
Comment (muesli) [inline verify_test.go:173]: Let's remove redundant newlines.

File: verify_test.go [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+					errors = append(errors, p.Error)
+				}
+			}
+			if len(errors) != tt.NumberOfExpectedErrors {
+				t.Errorf("%d errors were expected but %d occured", tt.NumberOfExpectedErrors, len(errors))
+			}
+		}
+
+	}
+
```
Comment (muesli) [inline verify_test.go:166]: Let's remove redundant newlines.

File: verify_test.go [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+func TestVerifySnapshot(t *testing.T) {
+	testPassword := "this_is_a_password"
+
+	for _, tt := range verifyTestCases {
+
```
Comment (muesli) [inline verify_test.go:279]: Let's remove redundant newlines.

File: verify.go [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+ */
+
+package knoxite
+
+import (
+	"math"
+	"math/rand"
+)
+
+func VerifyRepo(repository Repository, Percentage int) (prog chan Progress, err error) {
```
Comment (muesli) [inline verify.go:15]: Being a regular parameter here, we should use the lowercase version `percentage`.

File: verify.go [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+			prog <- p
+
+		}
+		close(prog)
+	}()
+
+	return prog, nil
+}
+
+func VerifyVolume(repository Repository, volumeId string, Percentage int) (prog chan Progress, err error) {
```
Comment (muesli) [inline verify.go:79]: Being a regular parameter here, we should use the lowercase version `percentage`.

File: verify.go [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+			prog <- p
+
+		}
+		close(prog)
+	}()
+
+	return prog, nil
+}
+
+func VerifySnapshot(repository Repository, snapshotId string, Percentage int) (prog chan Progress, err error) {
```
Comment (muesli) [inline verify.go:145]: Being a regular parameter here, we should use the lowercase version `percentage`.

File: verify_test.go [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+var verifyTestCases = []struct {
+	ErrorFunction          func(dir string) error
+	NumberOfExpectedErrors int
+	Percentage             int
+}{
+	{func(dir string) error { return nil }, 0, 100}, // Testing various Percentages
+	{func(dir string) error { return nil }, 0, 70},
+	{func(dir string) error { return nil }, 0, 30},
+	{func(dir string) error { return nil }, 0, 0},
+	{func(dir string) error { return nil }, 0, 256},
```
Comment (muesli) [inline verify_test.go:20]: I wonder if the last two cases (0% and 256%) shouldn't actually return errors instead.

File: verify_test.go [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+					errors = append(errors, p.Error)
+				}
+			}
+			if len(errors) != tt.NumberOfExpectedErrors {
+				t.Errorf("%d errors were expected but %d occured", tt.NumberOfExpectedErrors, len(errors))
+			}
+		}
+
+	}
+
```
Comment (muesli) [inline verify_test.go:272]: Let's remove redundant newlines.

File: verify_test.go [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+			for p := range progress {
+				if p.Error != nil {
+					errors = append(errors, p.Error)
+				}
+			}
+			if len(errors) != tt.NumberOfExpectedErrors {
+				t.Errorf("%d errors were expected but %d occured", tt.NumberOfExpectedErrors, len(errors))
+			}
+		}
+
```
Comment (muesli) [inline verify_test.go:376]: Let's remove redundant newlines.

**Pull request merged**